[WIP] OH integration

gedankenstuecke opened the pull request:
General API calls to make a connection on our end now works. Some to-dos before merging:

- [x] Fix upload of our details to Open Humans
- [x] check whether file on OH needs to be deleted first.
- [x] Make nicer views
- [x] Make Hound & CodeClimate happy
- [x] Add graphics to user profiles and `/openhumans/`.
- [x] Add connection to OH as phenotype to our data dumps. 
- [x] Write tests

gedankenstuecke commented:
Does Travis fail because I didn't update the `db/structure.sql` in the repo? @tsujigiri

tsujigiri commented on app/controllers/open_humans_profiles_controller.rb:
```diff
@@ -0,0 +1,101 @@
+    @oh_profile.save
+    upload_opensnp_id
+
+  end
```
You don't need instance variables if you don't use them outside of the method. Also, since this is OAuth, why not use one of the numerous OAuth client libs instead of hand-rolling it? Also... 😉 It's good practice to put most of the logic of, like, everything in [service classes](https://hackernoon.com/service-objects-in-ruby-on-rails-and-you-79ca8a1c946e) instead of stuffing the controllers like a christmas goose. 🦆

gedankenstuecke replied:
Do you have a recommendation for an OAuth client lib? The ones I saw while googling all felt a bit weird to me somehow, that's why I ended up writing it myself. 

But the service classes are a good idea. I'll refactor accordingly 👍

tsujigiri replied:
I haven't looked into it for a while, so I'm not sure. It just seems like re-inventing the wheel.

tsujigiri commented on app/models/open_humans_profile.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class OpenHumansProfile < ActiveRecord::Base
```
For that we'd need Rails5, first... 🐩

gedankenstuecke replied:
What exactly do you mean here? Follows the same convention as all other models?

tsujigiri replied:
Rubocop wants the models to subclass a comon `ApplicationRecord` class. I tried doing this in another project using Rails 4 and it didn't work. But, thinking about it, this may have been due to some gem we are using. Long story short, just ignore it for now. We can do this separately.

gedankenstuecke replied:
Ok!

tsujigiri commented:
@gedankenstuecke Yes, you need to commit the changed `db/structure.sql`. There's also [this](https://travis-ci.org/openSNP/snpr/builds/301765098#L677-L678), but I don't know if it's a problem.

gedankenstuecke commented on app/services/open_humans_service.rb:
```diff
@@ -0,0 +1,116 @@
+    metadata = generate_json(user)
+    metadata['tags'] = ['opensnp']
+    metadata['description'] = "links to openSNP user #{user.id}"
+    return metadata
```
this is actually not redundant as it otherwise breaks the hash…

tsujigiri replied:
This seems to be pretty redundant, as it is the last expression in the method. Am I missing something?

gedankenstuecke replied:
I don't know. If I'm not running the `return` it only returns `"links to openSNP user #{user.id}"` instead of the whole hash. That's why I put it back in, otherwise the OH API doesn't accept our upload.

tsujigiri replied:
Aaaaah! Now I get it. 💡  You don't need the `return`, but you still need the `metadata`. Without the `metadata` in the end, it returns the return value of the assignment above it, which is the value you assign.

gedankenstuecke commented on app/services/open_humans_service.rb:
```diff
@@ -0,0 +1,115 @@
+class OpenHumansService
```
is fixed…

gedankenstuecke commented on app/services/open_humans_service.rb:
```diff
@@ -0,0 +1,118 @@
+    post_body << "Content-Disposition: form-data; name=\"metadata\"\r\n\r\n"
+    post_body << metadata.to_json
+    post_body << "\r\n--#{boundary}\r\n"
+    post_body << "Content-Disposition: form-data; "
```
breaks the form if i do that…

gedankenstuecke commented on app/services/open_humans_service.rb:
```diff
@@ -0,0 +1,118 @@
+    post_body << "Content-Disposition: form-data; name=\"metadata\"\r\n\r\n"
+    post_body << metadata.to_json
+    post_body << "\r\n--#{boundary}\r\n"
+    post_body << "Content-Disposition: form-data; "
```
breaks the form if i do that…

tsujigiri replied:
Are you sure this is the only format the API accepts? This is usually what the browser generates from a form. If so, maybe you could put the whole thing in some form of template or a heredoc:

```ruby
def generate_form_body(user, boundary)
  metadata = generate_metadata(user)
  file_content = generate_json(user)
  <<-POST_BODY.strip_heredoc
    --#{boundary}\r
    Content-Disposition: form-data; name="project_member_id"\r\n\r
    #{user.open_humans_profile.project_member_id}\r
    --#{boundary}\r
    Content-Disposition: form-data; name="metadata"\r\n\r
    #{metadata.to_json}\r
    --#{boundary}\r
    Content-Disposition: form-data; name="data_file"; filename="#{user.id}.json"\r\n\r
    #{file_content.to_json}\r
    --#{boundary}--\r\n\r
  POST_BODY
end
```

gedankenstuecke replied:
you're right on this, it works! Thanks! 😍

gedankenstuecke commented:
Okay, the hound is basically happy now. I also followed the advice of @tsujigiri and put all the logic into an `OpenHumansService`, which makes the controller much nicer to look at. 

I also figured out how to avoid duplicate data entries on side of @OpenHumans. Slowly getting there 😄

tsujigiri commented on app/services/open_humans_service.rb:
```diff
@@ -1,94 +1,88 @@
     http.request(upload)
   end
 
   def delete_opensnp_id(user)
```
Since you use the `user` all over the place, how about passing it into the initializer and keeping it as an instance variable?

tsujigiri commented on app/services/open_humans_service.rb:
```diff
@@ -0,0 +1,111 @@
+  def upload_opensnp_id(user)
+    base_url = 'https://www.openhumans.org/api/direct-sharing/project/files/upload/?access_token='
+    uri = URI.parse(base_url + user.open_humans_profile.access_token)
+    boundary = '0P3NSNPH34RT50PENHUM4N5'
```
❤️

tsujigiri replied:
You could put that into a constant, too.

gedankenstuecke replied:
done :)

gedankenstuecke commented on app/services/open_humans_service.rb:
```diff
@@ -1,94 +1,88 @@
 
   def delete_opensnp_id(user)
-    url = URI.parse("https://www.openhumans.org/api/direct-sharing/project/files/delete/?access_token=#{user.open_humans_profile.access_token}")
+    base_url = 'https://www.openhumans.org/api/direct-sharing/project/files/delete/?access_token='
```
done :)

gedankenstuecke commented:
I think all of @tsujigiri's comments are addressed so far. Also hound & codeclimate are largely happy about the status now. I haven't written any tests so far. @tsujigiri any recommendations on how to best test an API integration?

gedankenstuecke commented:
Last update for (what is my) day: The views are finished now as well in principle:

* `/openhumans` lists users w/ connection and how to connect your own account if you haven't generated a connection so far
* the individual user-pages now have a small Open Humans logo badge, similar to the logo we put up for `fitbit` before. 
* the general menus include links to the index Open Humans index page, analogous to `fitbit` earlier. 

Still missing:
* including the link in our datadumps
* tests (i'd still appreciate input on how to best test for API calls)

tsujigiri commented on app/controllers/open_humans_profiles_controller.rb:
```diff
@@ -15,20 +23,14 @@ def authorize
+    oh_service = OpenHumansService.new(@user)
+    oh_service.authenticate(@code)
+    flash[:achievement] = 'Connected your account to Open Humans'
+    redirect_to @user
```
No instance variables needed if there is no view to use them in.

tsujigiri commented on app/services/open_humans_service.rb:
```diff
@@ -2,110 +2,120 @@
                       'code' => code,
-                      'redirect_uri' => 'http://localhost:3000/openhumans/authorize')
-    # set up request to use https
+                      'redirect_uri' => OH_REDIRECT_URL)
```
You can probably use `openhumans_authorize_path` here, right?

gedankenstuecke replied:
Mh… I don't exactly get what you mean

tsujigiri replied:
`OH_REDIRECT_URL` contains the full URL to `/openhumans/authorize`, which is what the URL helpers give you, which are generated from the routes. `openhumans_authorize_path` gives you the relative URL, `openhumans_authorize_url` the absolute one. For the latter to work, you need to configure the domain via the `default_url_options` in the environments:

```ruby
config.action_controller.default_url_options = { host: 'myproduction.com' }
```

gedankenstuecke commented on app/workers/zipfulldata.rb:
```diff
@@ -82,7 +82,12 @@ def create_user_csv(genotypes, zipfile)
       genotypes.each do |genotype|
         user = genotype.user
-        row = [user.id, genotype.fs_filename, user.yearofbirth, user.sex]
+        if user.open_humans_profile.nil?
```
I can't figure out what it wants of me here.

tsujigiri replied:
It wants you to use the return value of the whole `if` construct, to assign to `oh_name`:
```ruby
oh_name = if user.open_humans_profile.nil?
            '-'
          else
            user.open_humans_profile.open_humans_user_id
          end
```
... which can also be shortened to:

```ruby
oh_name = user.open_humans_profile&.open_humans_user_id || '-'
```

philippbayer replied:
Once this is fixed it can be merged, GREAT work!!

gedankenstuecke commented:
Ok, and some more updates now: 

* removed all the fitbit logic that would allow adding/editing fitbit as a data source. Like this we only have to delete the last two views after the grace period. 
* fixed the tests so that the new data dumps should contain the openhumans IDs
* added the Open Humans link to our data dumps

In principle it should now be good for merging  👍

tsujigiri commented on app/controllers/open_humans_profiles_controller.rb:
```diff
@@ -0,0 +1,42 @@
+    # does the user have an OH profile on openSNP? if not, create one
+    if user.open_humans_profile.nil?
+      user.open_humans_profile = OpenHumansProfile.new
+    end
```
I was confused because an `OpenHumansProfile` is assigned but not saved to the database, which happens later in the `OpenHumansService`. Maybe you move the assignment into the service as well. Seems to be a good place to do that.

tsujigiri commented:
Regarding testing, best is probably to write a feature spec using VCR to record the external calls. This can be a bit tricky with OAuth, but should be possible. Take a look at the other specs in `spec/features` for inspiration. 🙂

tsujigiri commented:
I think it would be good to move the fitbit-related things to another pull-request, to not let this one grow more than necessary.

gedankenstuecke commented:
@tsujigiri thanks for the comments! I fixed the remaining issue as you suggested and will have a look at the tests. 

Re the fitbit parts: I'd personally keep it in this one as it should be deployed alongside?

tsujigiri commented:
@gedankenstuecke I'm just thinking, what if you want to roll back the OpenHumans integration, but not the removal of the Fitbit integration? Makes it easier if they are separate. Not that I anticipate that. 😉

gedankenstuecke commented:
That’s a good point. What would be the best way to achieve that?

tsujigiri commented:
Since it's a separate commit, you could create a new branch from this one and do an interactive rebase on each branch, removing the commits belonging to the respective other branch.

Something along the lines of:

```sh
# Create new branch from this one.
git checkout oh-integration -b remove-fitbit
# Do an interactive rebase. This opens an editor, just remove the commits you want to remove.
git rebase -i master...
# The same on this branch.
git checkout oh-integration
git rebase -i master...
```

The `master...` refers to the range of commits in the current branch that are not on the master branch.

gedankenstuecke commented:
Actually it's not that clean a commit unfortunately. Splitting this up would be a bit more work. How about I keep that in mind for next time though? :D

tsujigiri commented:
Sure. 👍
